File: leetcode-20/cpp/leetcode-20.cpp
#include "test.h"
#include <iostream>
#include <stack>
#include <string>

class Solution {
public:
	bool isValid(std::string s) {
		if (s.length() % 2 != 0) {
			std::cout << "s.length = " << s.length() % 2;
			return 0;
		}

		if (s.empty()) {
			return true;
		}
		
		std::stack<char> parn;
		for (auto c : s) {
			switch (c){
			case '(':
			case '[':
			case '{':{
				parn.push(c);
				break;
			}
			case ')': {
				if (parn.empty() || parn.top() != '(') {
					return false;
				}
				parn.pop();
				break;
			}
			case ']': {
				if (parn.empty() || parn.top() != '[') {
					return false;
				}
				parn.pop();
				break;
			}
			case '}': {
				if (parn.empty() || parn.top() != '{') {
					return false;
				}
				parn.pop();
				break;
			}
			default:
				break;
			}
		}

		return parn.empty();
	}
};

bool it1(std::string param, bool expect) {
	Solution s;
	bool result = s.isValid(param);

	if (result == expect) {
		std::cout << "check " << param << " successed." << std::endl;
	}
	else {
		std::cout << "check " << param << " failed. expect " << expect << " but result " << result << std::endl;
	}
	return expect == result;
}

bool describe20() {
	it1("()", true);
	it1("()[]{}", true);
	it1("(]", false);
	it1("([)]", false);
	it1("{[]}", true);
	return true;
}

File: leetcode-20/python/leetcode-20.py
class Solution:
    def isValid(self, s):
        """

        :param s: str
        :return:  bool
        """
        if len(s) == 0:
            return True
        if len(s) % 2 != 0:
            return False

        stack = []

        for c in list(s):
            print(c)
            if c == '(' or c == '[' or c == '{' :
                stack.append(c)
                continue
            if c == ')':
                if len(stack) == 0 or stack[len(stack) - 1] != '(':
                    return False
            if c == ']':
                if len(stack) == 0 or stack[len(stack) - 1] != '[':
                    return False
            if c == '}':
                if len(stack) == 0 or stack[len(stack) - 1] != '{':
                    return False
            stack.pop()

        return len(stack) == 0


if __name__ == '__main__':
    s = Solution()

    result = s.isValid('({})')
    print(result)

    result = s.isValid('()[]{}')
    print(result)

    result = s.isValid('(]')
    print(result)

    result = s.isValid('')
    print(result)

    result = s.isValid('(){')
    print(result)
